### ISSUE TITLE
improve performance of json formatter

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#35501]
*Title*: *improve performance of json formatter*

*Description*:

See the performance result:

```
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
BM_AccessLogFormatterSetup             8899 ns         8851 ns        78731
BM_AccessLogFormatter                   486 ns          485 ns      1452972
BM_StructAccessLogFormatter            7433 ns         7423 ns        92404
BM_TypedStructAccessLogFormatter       5459 ns         5394 ns       124266
BM_JsonAccessLogFormatter             12378 ns        12364 ns        52029
BM_TypedJsonAccessLogFormatter        11297 ns        11282 ns        65591
BM_FormatterCommandParsing             40.5 ns         40.5 ns     17290397
```

The json formatter takes 20+x overhead rather than text formatter. What can I say? I always know it is slow, but I don't know it's so slow.

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

**Comment by wbpcode** [comment]
cc @jmarantz

**Comment by jmarantz** [comment]
Try using the json streamer I wrote: https://github.com/envoyproxy/envoy/blob/main/source/common/json/json_streamer.h

This was a dramatic speedup for json stats.

**Comment by wbpcode** [comment]
Yeah, it's great for constructing a json string. But I think it's not appropriate for this scenario. See, formatting the output proto struct has already taken 10+x times than the text formatter.

What I can think of is try to do as much work as possible when loading the configuration. For example, we can load the json formatter configuration to a list of raw string and handler. And at runtime, we can iterate the list and join the raw string and output of handler. By this way, the json formatter may could provider similar performance with the text formatter. (I need to have a quick try first.)

For example:
```
json_format:
  key: %COMMAND%

->

'{"key":' + '<command handler>' + '}'
```

**Comment by jmarantz** [comment]
i'm not sure why it wouldn't be appropriate for this scenario but I"m not famiilar enough with the existing code. It does involve re-thinking, to some extent, the approach to generating json.

In particular, the anti-pattern (IMO, from a performance perspective), is converting bulk text into a protobuf as an intermediate format. We should try to avoid making intermediate copies of anything. That is the key to making json converters fast.

**Comment by kyessenov** [comment]
@wbpcode approach should work - if we can avoid the joining at runtime and not escape the values, it'll probably be the same performance as text.

**Comment by wbpcode** [comment]
> i'm not sure why it wouldn't be appropriate for this scenario

After a quick try, I think it could works, as long as we don't convert the output to protobuf. We can parse the json format configuration to a tree that composite of raw string and formatter providers. Then we can iterate the tree at runtime and use the json streamer to construct the output json. This should pretty faster than current implementation.

But I still select another way. I will parse the configuration and get a tree, then I will iterate the tree at configuration loading time directly and generate an array of raw string and formatter providers. Adjacent raw strings will be merged to one when doing the iteration.
Then, at runtime, we only need to iterate the array. This way provide the best performance. Here is a result. It provides comparable performance with raw text formatter and is 16-25x faster than the original json formatter.

```
----------------------------------------------------------------------------
Benchmark                                  Time             CPU   Iterations
----------------------------------------------------------------------------
BM_AccessLogFormatter                    492 ns          492 ns      1390102
BM_StructAccessLogFormatter             6846 ns         6838 ns        97314
BM_TypedStructAccessLogFormatter        5907 ns         5900 ns       118040
BM_JsonAccessLogFormatter              12761 ns        12747 ns        55039
BM_TypedJsonAccessLogFormatter         11589 ns        11572 ns        60475
BM_JsonAccessLogFormatterNew             512 ns          511 ns      1381488
BM_TypedJsonAccessLogFormatterNew        706 ns          705 ns       949938
```

**Comment by wbpcode** [comment]
> In particular, the anti-pattern (IMO, from a performance perspective), is converting bulk text into a protobuf as an intermediate format. We should try to avoid making intermediate copies of anything. That is the key to making json converters fast.

Yeah. Our expected final result is string. But use the proto struct or proto value as intermediate output. It bring much overhead.

**Comment by sathyaj01** [comment]
We observed a similar performance issue with the Json log formatter. A recent session of CPU profiling revealed that a significant portion of CPU usage was attributed to the Json log formatter.
Below is the CPU profiling with Json log formatter setup.
 
![image](https://github.com/user-attachments/assets/08453026-10a4-4e41-9e24-d4a171074cfd)

CPU profiling with text formatter.
 
![image](https://github.com/user-attachments/assets/884b9ded-2c21-4a06-97c5-7d48baa4dc43)

We provided similar traffic for both profiling sessions, averaging around 2,000 QPS (queries per second). While CPU usage can vary depending on the type of request, in this profiling, we used consistent traffic for both sessions.
Here is a comparison of Json and Text formatter in terms of queries per second.

 
![image](https://github.com/user-attachments/assets/678139f8-5628-4a8d-b9d4-ab2451b07fa1)
